### PR TITLE
feat: Use networks instead of currencies

### PR DESCRIPTION
**Pull request by lszetg**
# Pull Request Template

## Motivation and Context

It's a continuation of the ticket on the SDK side.

## Summary

- `set_currency` method has been removed
- `get_networks` and `set_network` methods have been added
- api has been adjusted to work with the backend
- `node_urls` has been removed from `Config` (`node_url` is now part of `Network`)

## Checklist

Please ensure that your PR meets the following requirements:

### General

- [ ] Code follows project style and guidelines.
- [ ] No redundant or duplicate code.
- [ ] No added new dependencies without clear justification and approval.

### Documentation

- [ ] Added/updated relevant documentation (e.g., README, inline comments).
- [ ] Added a CHANGELOG entry for major changes.

### Testing

- [ ] Added relevant test cases, leveraging tools like `rstest` or similar for reusable tests.
- [ ] Tests pass locally.

### Build & CI/CD

- [ ] Confirmed no CI/CD pipeline issues.
- [ ] Updated build scripts where necessary.

## Comments to Reviewer

Please provide any comments or points of consideration for the reviewers, such as specific areas of focus, potential edge cases, or context that would aid the review process.

### Reviewer Responsibilities

- [ ] Code adheres to style and guidelines.
- [ ] All tests are appropriately covered and they pass.
- [ ] Consider performance and/or security impacts.
- [ ] Constructive feedback is provided.

### CODE REVIEW

File: sdk/src/backend/user.rs [hunk 1]
```diff
@@ -131,18 +126,18 @@ pub async fn set_preferred_currency(
...
     config: &Config,
     access_token: &AccessToken,
     username: &str,
-) -> Result<Option<Currency>> {
+) -> Result<Option<String>> {
     let base_url = &config.backend_url;
-    let url = format!("{base_url}/user/currency");
+    let url = format!("{base_url}/user/network");
     info!("Used url: {url:#?}");
     info!("Getting preferred currency for {username}");
```
Comment (lszetg) [inline sdk/src/backend/user.rs:142]: currency => network

File: sdk/src/core/transaction.rs [hunk 2]
```diff
@@ -326,7 +317,7 @@ impl Sdk {
                         receiver: val.outgoing.username,
                         reference_id: val.index,
                         amount: val.incoming.amount.try_into()?,
-                        currency: val.incoming.currency.to_string(),
+                        currency: val.incoming.network.currency, // ?????????????????????
```
Comment (lszetg) [inline sdk/src/core/transaction.rs:320]: remove ??????

File: sdk/src/wallet/error.rs [hunk 3]
```diff
@@ -49,10 +45,17 @@ pub enum WalletError {
...
     MissingNodeUrls {
-        /// the currency for which node url is missing
-        currency: Currency,
+        /// the id of the network for which node url is missing
+        network_id: String,
+    },
+
+    /// Error raises if network is missing in config
+    #[error("Network not present in config {network_id:?}")]
+    NetworkNotPresentInConfig {
```
Comment (lszetg) [inline sdk/src/wallet/error.rs:56]: is it used?

File: sdk/bindings/android/src/lib.rs [hunk 4]
```diff
@@ -88,9 +88,6 @@ mod ffi {
     ///     "backend_url": "<valid URL to the backend API>",
     ///     "storage_path": "/path/to/valid/folder",
     ///     "log_level": "info",
-    ///     "node_urls": {
```
Comment (lszetg) [inline sdk/bindings/android/src/lib.rs:91]: support multiple networks

File: sdk/bindings/android/src/lib.rs [hunk 5]
```diff
@@ -107,26 +104,22 @@ mod ffi {
...
     /// @return Serialized string of a hashmap with currencies as key and node urls as value
-    pub fn getNodeUrls() -> Result<String, String> {
+    pub fn getNetworks() -> Result<String, String> {
         let result = runtime().block_on(async move {
             let sdk = get_or_init_sdk().write().await;
-            sdk.get_node_urls_backend().await
+            sdk.get_networks()
         });
 
-        match result {
```
Comment (lszetg) [inline sdk/bindings/android/src/lib.rs:116]: fix it

File: sdk/bindings/swift/src/ffi_functions.rs [hunk 6]
```diff
@@ -1285,33 +1279,34 @@ impl CawaenaSdk {
...
-    /// Get the user's preferred currency.
+    /// Get the user's preferred network.
     ///
     /// # Returns
     ///
-    /// * Ok - the preferred currency, or `None` if it has not been set.
+    /// * Ok - the preferred network, or `None` if it has not been set.
     /// * Err - if there was an error contacting the backend.
-    pub async fn get_preferred_currency(&self) -> Result<PreferredCurrency, String> {
+    pub async fn get_preferred_network(&self) -> Result<String, String> {
```
Comment (lszetg) [inline sdk/bindings/swift/src/ffi_functions.rs:1288]: should be `Result<Option<String>, String>`

File: sdk/examples/utils.rs [hunk 7]
```diff
@@ -46,7 +36,9 @@ pub async fn init_sdk() -> (Sdk, CleanUp) {
     let access_token = AccessToken::try_from(access_token).unwrap();
     sdk.refresh_access_token(Some(access_token)).await.unwrap();
 
-    sdk.set_currency(Currency::Iota);
+    sdk.set_network(String::from("67a1f08edf55756bae21e7eb"))
```
Comment (lszetg) [inline sdk/examples/utils.rs:39]: replace hardcoded id

File: sdk/src/core/config.rs [hunk 8]
```diff
@@ -164,8 +136,18 @@ impl Sdk {
         Ok(())
     }
 
-    /// Get supported currencies and node urls from backend
-    pub async fn get_node_urls_backend(&self) -> Result<HashMap<String, Vec<String>>> {
+    /// Set networks
+    pub fn set_networks(&mut self, networks: Option<Vec<Network>>) {
```
Comment (lszetg) [inline sdk/src/core/config.rs:140]: `pub` -> `pub(crate)`

File: sdk/src/core/config.rs [hunk 9]
```diff
@@ -164,8 +136,18 @@ impl Sdk {
...
+        self.networks = networks;
+    }
+
+    /// Get networks
+    pub fn get_networks(&self) -> Option<Vec<Network>> {
+        self.networks.clone()
+    }
+
+    /// Get supported networks from backend
+    pub async fn get_networks_backend(&self) -> Result<Vec<Network>> {
```
Comment (lszetg) [inline sdk/src/core/config.rs:150]: `pub` -> `pub(crate)`

**Review by hauju — APPROVED** [review]

File: sdk/src/core/config.rs (hunk 8 above)
Comment (lszetg) [reply sdk/src/core/config.rs:140]: image
cannot do it, could be there is some better design that we can apply :D

File: sdk/examples/utils.rs (hunk 7 above)
Comment (lszetg) [reply sdk/examples/utils.rs:39]: i've tried, cannot reach to the function inside `testing_utils`

File: sdk/bindings/android/src/lib.rs (hunk 5 above)
Comment (lszetg) [reply sdk/bindings/android/src/lib.rs:116]: cannot apply it because `get_backend` takes current (loaded) list of networks
created a ticked to make it query the backend, then it can be reverted
- https://jira.etogruppe.com/browse/ZD0006-571

**Review by antbern — APPROVED** [review]
Let's do this! 🚀 
Btw, you need to sign your commits to be able to merge them @lszetg . A bit frustrating but better for security 🌵

File: sdk/src/core/config.rs (hunk 8 above)
Comment (antbern) [reply sdk/src/core/config.rs:140]: Alright, well leave it as-is for now and we can look at streamlining this later 💯

File: sdk/bindings/android/src/lib.rs (hunk 5 above)
Comment (antbern) [reply sdk/bindings/android/src/lib.rs:116]: Ah yes it does not return a `Result`, but rather an `Option`. What if we just do
```rust
sdk.get_networks().unwrap_or_default()
```
to just default to an empty `Vec` if there are no networks, and then the `serde_json::to_string` can stay?

**Review by antbern — APPROVED** [review]
